Tests/PackageLoading/InMemoryConventionTests.swift

```diff
@@ -0,0 +1,174 @@
+        }
+        result(self)
+    }
+}
+
+extension PackageModel.Package {
+    var allModules: [Module] {
+        return modules + testModules
+    }
+}
+
+enum PackageBuilderTester {
+    case package(PackageModel.Package)
+    case error(Swift.Error)
+
+    func checkError(_ str: String) {
+        switch self {
+        case .package(let package):
+            XCTFail("\(package) did not have any error")
+        case .error(let error):
+            XCTAssertEqual(String(error), str)
+        }
+    }
+
+    func checkModule(_ name: String, result: ((ModuleResult) -> Void)? = nil) {
+        guard case .package(let package) = self else {
+            return XCTFail("Error: \(self)")
+        }
+        guard let module = package.allModules.filter({$0.name == name}).first else {
+            return XCTFail("Module: \(name) not found")
+        }
+        result?(ModuleResult(module))
+    }
+
+    final class ModuleResult {
+        let module: Module
+        lazy var sources: Set<RelativePath> = { Set(self.module.sources.relativePaths) }()
+
+        init(_ module: Module) {
+            self.module = module
+        }
+
+        func check(c99name: String) {
+            XCTAssertEqual(module.c99name, c99name)
+        }
+
+        func check(type: ModuleType) {
+            XCTAssertEqual(module.type, type)
+        }
+
+        func check(isTest: Bool) {
+            XCTAssertEqual(module.isTest, isTest)
+        }
+
+        func checkSources(root: AbsolutePath) {
+            XCTAssertEqual(module.sources.root, root)
+        }
+
+        func checkSources(count: Int) {
+            XCTAssertEqual(module.sources.paths.count, count)
+        }
+
+        func checkSources(_ paths: RelativePath...) {
+            for path in paths {
+                XCTAssert(sources.contains(path), "\(path.asString) not found in module \(module.name)")
+            }
+        }
+    }
+}
+
+/// Tests for the handling of source layout conventions.
+class InMemoryConventionTests: XCTestCase {
+
+    func testMixedSources() throws {
+        var fs = InMemoryFileSystem()
+        try fs.createEmptyFiles("/home/MyPackage/Sources/main.swift",
+                                "/home/MyPackage/Sources/main.c")
+        PackageBuilderTester("/home/MyPackage", in: fs) { result in
+            result.checkError("the module at /home/MyPackage/Sources contains mixed language source files fix: use only a single language within a module")
+        }
+    }
+
+    func test1() throws {
+        var fs = InMemoryFileSystem()
+        try fs.createEmptyFiles("/home/MyPackage/Sources/ModuleA/main.swift",
+                                "/home/MyPackage/Sources/ModuleB/main.c",
+                                "/home/MyPackage/Sources/ModuleB/foo.c")
+
+        PackageBuilderTester("/home/MyPackage", in: fs) { result in
+            result.checkModule("ModuleA") { moduleResult in
+                moduleResult.check(c99name: "ModuleA")
```

Comment:
I thought about it but it forces clients to check for all these things even if they want to but that's probably a good thing and less LOC